lib/models/requestResponse.js

```diff
@@ -1,12 +1,116 @@
 'use strict';
 
-const
-  assert = require('../utils/assertType');
+const assert = require('../utils/assertType');
 
 const
   _request = 'request\u200b',
   _headers = 'headers\u200b';
 
+class Headers {
+  constructor() {
+    this.namesMap = new Map();
+    this.headers = {};
+    this.proxy = new Proxy(this.headers, {
+      get: (target, name) => this.getHeader(name),
+      set: (target, name, value) => this.setHeader(name, value),
+      deleteProperty: (target, name) => this.removeHeader(name)
+    });
+  }
+
+  getHeader (name) {
+    if (typeof name === 'symbol') {
+      return this.headers[name];
+    }
+
+    assert.assertString('header name', name);
+
+    if (!name) {
+      return;
+    }
+
+    return this.headers[this.namesMap.get(name.toLowerCase())];
+  }
+
+  removeHeader (name) {
+    assert.assertString('header name', name);
+
+    if (!name) {
+      return true;
+    }
+
+    const
+      lowerCased = name.toLowerCase(),
+      storedName = this.namesMap.get(lowerCased);
+
+    if (storedName) {
+      delete this.headers[storedName];
+      this.namesMap.delete(lowerCased);
+    }
+
+    return true;
+  }
+
+  setHeader (name, value) {
+    assert.assertString('header name', name);
+
+    if (!name) {
+      return true;
+    }
+
+    const
+      lowerCased = name.toLowerCase(),
+      _value = String(value);
+
+    let _name = this.namesMap.get(lowerCased);
+
+    if (!_name) {
+      this.namesMap.set(lowerCased, name);
+      _name = name;
+    }
+
+    // Common HTTP headers are overwritten when set, instead of being
+    // concatenated
+    switch (lowerCased) {
```

Comment:
I'm a little bit uncomfortable with this list, how can we ensure that we do not forget important headers ?

Reply:
This is not meant to be exhaustive, it's just the most frequent HTTP headers that we know holds only 1 value. 
The original version of this setter always concatenated new values to existing headers, and this was upsetting our clients working with HTTP, so we added this list to make things a bit easier.

I guess that in a future major version, we can get rid of it completely by:
* never concatenating new values to existing ones (diverging from the behavior of Node.js' response.setHeader)
* allowing any kind of values, not just string ones, so that users can store arrays, and stringifying header values just before they are sent to the network